### PR TITLE
Efficiency improvement

### PR DESCRIPTION
**Pull request by mgjarrett**
## Description

The uniform mesh converter can take a very long time. Profiling the converter identified a couple of bottlenecks that turned out to be relatively low-hanging fruit. This is partially documented in #991 and #1039 .

These updates reduced uniform mesh conversion time by 80% by avoiding the `copy.deepcopy(block)`. This reduces the amount of detail stored on the uniform mesh reactor, but it is still sufficient for the physics that need to be performed on the uniform mesh reactor.

These updates also reduced operator initialization time by about 40% when thermal expansion is enabled.

---

## Checklist

- [x] This PR has only one purpose or idea.
- [x] Tests have been added/updated to verify that the new/changed code works.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any bug fixes or new features.
- [x] The documentation is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `setup.py`.

### CODE REVIEW

File: armi/reactor/blocks.py [hunk 1]
```diff
@@ -157,6 +158,47 @@ def __deepcopy__(self, memo):
 
         return b
 
+    def createBlankCopy(self):
```
Comment (jakehader) [inline armi/reactor/blocks.py:161]: This might be better as a function somewhere else because this adds design details into the generic block class. This could be closer to where it's applied as a function for now and upgraded to a method later maybe.

File: armi/reactor/composites.py [hunk 2]
```diff
@@ -1249,12 +1249,14 @@ def getNuclideNumberDensities(self, nucNames):
...
 
-        nucDensForEachComp = numpy.array(
-            [
-                [c.getNumberDensity(nuc) for nuc in nucNames]
-                for c in self.iterComponents()
-            ]
-        )  # c x n
+        densListForEachComp = []
+        for c in self.iterComponents():
+            numberDensityDict = c.getNumberDensities()
```
Comment (jakehader) [inline armi/reactor/composites.py:1254]: This is faster because of using `getNumberDensities()`, right?

File: armi/reactor/converters/uniformMesh.py [hunk 3]
```diff
@@ -581,14 +604,17 @@ def clearStateOnAssemblies(assems, blockParamNames=None, cache=True):
 
         cachedBlockParamData = collections.defaultdict(dict)
 
-        blocks = []
-        for a in assems:
-            blocks.extend(a.getBlocks())
-        for b in blocks:
+        if len(assems) > 0:
```
Comment (jakehader) [inline armi/reactor/converters/uniformMesh.py:607]: You might be able to save an indent if you do

```
if not assems:
   return cachedBlockParamData
```

File: armi/reactor/converters/uniformMesh.py [hunk 4]
```diff
@@ -853,48 +913,13 @@ def _setParamsToUpdate(self, direction):
                     if not name in excludedParamNames
                 ]
             )
+        if direction == "in":
+            self.blockParamNames.extend(["molesHmBOL", "massHmBOL"])
```
Comment (jakehader) [inline armi/reactor/converters/uniformMesh.py:917]: Should these be added to the parameter categories or why does it make sense to have this here?

File: armi/reactor/blocks.py (hunk 1 above)
Comment (mgjarrett) [reply armi/reactor/blocks.py:161]: This implementation did become specific to a `HexBlock`, but it's maybe a functionality that should exist for other types of blocks also.

File: armi/reactor/composites.py (hunk 2 above)
Comment (mgjarrett) [reply armi/reactor/composites.py:1254]: Yes. This is small compared to some of the other changes, but `composites.getNumberDensity()` calls 

* `getNuclideNumberDensities()` if the the `composite` is not a `component`, which then calls `getNumberDensities()` on the child components
*  `self.p.numberDensities.get(nucName, 0.0)` if it is a `component`


https://github.com/terrapower/armi/blob/15a710f0620ca9521370f11b828de352f7b939e4/armi/reactor/composites.py#L1234-L1235

File: armi/reactor/converters/uniformMesh.py [hunk 5]
```diff
@@ -415,18 +421,24 @@ def makeAssemWithUniformMesh(
                 sourceBlock = blocks[0]
                 xsType = blocks[0].p.xsType
 
-            block = copy.deepcopy(sourceBlock)
+            block = sourceBlock.createBlankCopy()
             block.p.xsType = xsType
             block.setHeight(topMeshPoint - bottom)
             block.p.axMesh = 1
+            block.p.zbottom = bottom
```
[the remark this answers is not in the document]
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:428]: It would if the `block` was added to `newAssem` before `block.setHeight()` is called. Maybe it should be rearranged to make that happen.

File: armi/reactor/converters/uniformMesh.py (hunk 4 above)
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:917]: They're already in the `depletion` category.

1. We could add `depletion` to the categories to be mapped, but then we'd also get a bunch of other parameters that we don't need.
2. We could create a new category `heavyMetalMass` or something along those lines and use it here. But I think we should avoid creating too many parameter categories for very specific use cases.

I'm trying to balance having a reasonable number of categories against having explicit parameter mapping specifications like this. If it was 10 parameters I would create a category, but for two parameters it seems better to just explicitly name them.

File: armi/materials/mixture.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+from armi import materials
+
+
+class Mixture(materials.Material):
+    """
+    Homogenized mixture of materials
+
+    .. warning:: This class is meant to be used for homogenized block models for neutronics and other
+       physics solvers.
+    """
```
Comment (john-science) [inline armi/materials/mixture.py:28]: It feels like we should mention this is her for performance reasons, and not physics one.  

Opinion?

File: armi/physics/neutronics/parameters.py [hunk 7]
```diff
@@ -579,6 +579,7 @@ def linPowByPinGamma(self, value):
     with pDefs.createBuilder(
         default=0.0,
         location=ParamLocation.AVERAGE,
+        categories=[parameters.Category.detailedAxialExpansion],
```
Comment (john-science) [inline armi/physics/neutronics/parameters.py:582]: Is this related to this PR, or just a bonus?

File: armi/reactor/blocks.py [hunk 8]
```diff
@@ -1519,6 +1533,58 @@ def coords(self, rotationDegreesCCW=0.0):
             round(y, units.FLOAT_DIMENSION_DECIMALS),
         )
 
+    def createCopy(self):
+        """
+        Create a blank copy of a block (ligher weight than deepcopy)
```
Comment (john-science) [inline armi/reactor/blocks.py:1538]: "lighter weight" might be more clear if it said "faster".  Just IMO. I won't fight for it.

File: armi/reactor/blocks.py [hunk 9]
```diff
@@ -1519,6 +1533,58 @@ def coords(self, rotationDegreesCCW=0.0):
...
+
+        b = self.__class__(self.getName(), height=self.getHeight())
+        b.setType(self.getType())
+
+        # assign macros and LFP
+        b.macros = self.macros
+        b._lumpedFissionProducts = self._lumpedFissionProducts
+        b.p.buGroup = self.p.buGroup
+
+        hexComponent = Hexagon(
```
Comment (john-science) [inline armi/reactor/blocks.py:1553]: Sorry, stupid question, does this work for non-hex blocks?

Because 2/3 of our users don't have hex blocks.

File: armi/reactor/converters/uniformMesh.py [hunk 10]
```diff
@@ -400,13 +405,13 @@ def makeAssemWithUniformMesh(
...
                     totalHeight = sum(typeHeight.values())
-                    runLog.extra(
+                    runLog.debug(
                         f"Multiple XS types exist between {bottom} and {topMeshPoint}. "
                         f"Using the XS type from the largest region, {xsType}"
                     )
                     for xs, h in typeHeight.items():
                         heightFrac = h / totalHeight
-                        runLog.extra(f"XSType {xs}: {heightFrac:.4f}")
+                        runLog.debug(f"XSType {xs}: {heightFrac:.4f}")
```
Comment (john-science) [inline armi/reactor/converters/uniformMesh.py:414]: I think we need more than this.

Just because a single log line was useful to someone 8 years ago, doesn't mean anyone has looked at it in the last 7 years.

And I _definitely_ think we have a lot of `debug` log lines masquerading as `extra` and even `info`.

Nice!

File: armi/reactor/converters/uniformMesh.py (hunk 4 above)
Comment (john-science) [reply armi/reactor/converters/uniformMesh.py:910]: Could we at least save those strings off as constants at the top of the file?

```python
HEAVY_METAL_PARAMS = ["molesHmBOL", "massHmBOL"]
```

Or some such.

File: armi/reactor/blocks.py [hunk 11]
```diff
@@ -157,6 +158,19 @@ def __deepcopy__(self, memo):
...
+    def createCopy(self):
+        """
+        Create a copy of a block
+
+        Used to implement a copy function for a specific block types that is lighter
+        than a deepcopy. This base class implementation is just a deepcopy.
+
+        Notes
+        -----
+        Only impelemented for HexBlock right now.
```
Comment (jakehader) [inline armi/reactor/blocks.py:170]: `implemented`

File: armi/reactor/blocks.py (hunk 9 above)
Comment (jakehader) [reply armi/reactor/blocks.py:1553]: I think that this is addressed on the general block as just a deepcopy now.

File: armi/materials/mixture.py (hunk 6 above)
Comment (mgjarrett) [reply armi/materials/mixture.py:28]: Yes, I will add more detail to this docstring.

File: armi/physics/neutronics/parameters.py (hunk 7 above)
Comment (mgjarrett) [reply armi/physics/neutronics/parameters.py:582]: It's part of this PR. Some of these reaction rate parameters need to be mapped when performing gamma transport using a downstream plugin.

Before this PR, the uniform mesh converter would `deepcopy` blocks from the source assembly to the destination assembly in `makeAssemWithUniformMesh`, and these parameters would have values acquired from the `deepcopy`. Now that we have replaced the `deepcopy` with a less comprehensive copy function, the parameters need to be explicitly mapped so that the values are present on the uniform mesh reactor.

File: armi/reactor/blocks.py (hunk 8 above)
Comment (mgjarrett) [reply armi/reactor/blocks.py]: I agree. I added detail to the docstring to make it clear what this is for and why it would be used.

I originally said "lighter weight" to indicate that the block created from this function has less data associated with it than one created using `deepcopy`. But yes, the main goal is for `createCopy` to be significantly faster than a full `deepcopy` of the block.

File: armi/reactor/blocks.py (hunk 9 above)
Comment (mgjarrett) [reply armi/reactor/blocks.py:1579]: That's correct @jhader. A faster/lighter copy function could be written for Cartesian or other blocks and it would probably look quite similar to this one, but for now I'm trying to minimize the scope of the change.

I don't have enough knowledge the use cases for other blocks to write a good `createCopy` that generalizes to all block types, but someone who uses another block type frequently should be able to implement a function modeled on this one without much trouble.

File: armi/reactor/converters/uniformMesh.py (hunk 10 above)
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:416]: Thanks! Unfortunately, I think I'm the one who initially put this in as an `extra` just a couple months ago. I didn't realize how much obnoxious output it creates until I started testing with `detailedAxialExpansion: true`

File: armi/reactor/blocks.py [hunk 12]
```diff
@@ -1519,6 +1536,84 @@ def coords(self, rotationDegreesCCW=0.0):
...
+        # only works if there is a single clad component that has a MultiIndexLocation
+        if self.hasComponents(Flags.CLAD):
+            pinComponent = Circle(
+                "voidPin",
+                "Void",
+                self.getAverageTempInC(),
+                self.getAverageTempInC(),
+                0.0,
+            )
+            clad = self.getComponent(Flags.CLAD)
```
Comment (keckler) [inline armi/reactor/blocks.py:1601]: This is causing an error for me.

I have blocks that have multiple components with the `CLAD` flag, because I have some assemblies that have multiple fuel types loaded into it.

Comment (keckler) [reply armi/reactor/blocks.py:1601]: I guess you note this in a comment above...

File: armi/reactor/blocks.py [hunk 13]
```diff
@@ -1587,27 +1587,29 @@ def _createHomogenizedCopy(self):
...
-            b.add(pinComponent)
+        if pinSpatialLocators:
+            # create a null component with cladding flags and spatialLocator from source block's
+            # clad component in case pin locations need to be known for physics solver
+            # only works if there is a single clad component that has a MultiIndexLocation
+            if self.hasComponents(Flags.CLAD):
+                cladComponents = self.getComponents(Flags.CLAD)
+                for i, clad in enumerate(cladComponents):
+                    pinComponent = Circle(
+                        "voidPin{i}",
```
Comment (keckler) [inline armi/reactor/blocks.py:1598]: Does this need to be an f-string?

File: armi/materials/mixture.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+# limitations under the License.
+
+"""
+Homogenized mixture material
+"""
+
+from armi import materials
+
+
+class Mixture(materials.Material):
```
[the remark this answers is not in the document]
Comment (mgjarrett) [reply armi/materials/mixture.py:22]: I would probably agree that we generally wouldn't want users to use this material. There could be a situation where a user has homogenized number densities for a material they want to model, but I guess the user could just create a custom material for that. Custom materials are a little outside my typical user experience with ARMI, so I would defer on this.

File: armi/reactor/blocks.py [hunk 15]
```diff
@@ -1587,27 +1587,29 @@ def _createHomogenizedCopy(self):
...
-                for i1, i2 in zip(
-                    list(pinComponent.spatialLocator), list(clad.spatialLocator)
-                ):
-                    i1.associate(i2.grid)
-            pinComponent.setDimension("mult", clad.getDimension("mult"))
-            b.add(pinComponent)
+        if pinSpatialLocators:
+            # create a null component with cladding flags and spatialLocator from source block's
+            # clad component in case pin locations need to be known for physics solver
+            # only works if there is a single clad component that has a MultiIndexLocation
```
Comment (keckler) [inline armi/reactor/blocks.py:1593]: Comment is out of date

File: armi/reactor/blockParameters.py [hunk 16]
```diff
@@ -93,14 +93,12 @@ def detailedNDens(self, value):
...
             units="mole",
             description="Total number of atoms of heavy metal at BOL assuming a full assembly",
-            default=None,
         )
 
         pb.defParam(
             "massHmBOL",
             units="grams",
             description="Mass of heavy metal at BOL",
-            default=None,
```
[the remark this answers is not in the document]
Comment (mgjarrett) [reply armi/reactor/blockParameters.py:103]: These are scalar values, so the default should be 0.0 rather than `None`. I don't have it documented, but I encountered errors with the defaults as `None`.

File: armi/reactor/blocks.py (hunk 13 above)
Comment (mgjarrett) [reply armi/reactor/blocks.py:1598]: yeah...

File: armi/reactor/blocks.py [hunk 17]
```diff
@@ -1519,6 +1535,85 @@ def coords(self, rotationDegreesCCW=0.0):
...
+        hexComponent = Hexagon(
+            "homogenizedHex",
+            "Mixture",
+            self.getAverageTempInC(),
+            self.getAverageTempInC(),
+            self._pitchDefiningComponent[1],
+        )
+        emptyNDens = {nuc: 0.0 for nuc in self.getNuclides()}
+        hexComponent.setNumberDensities(emptyNDens)
+        b.add(hexComponent)
```
[the remark this answers is not in the document]
Comment (mgjarrett) [reply armi/reactor/blocks.py]: I renamed it to `_createHomogenizedCopy()`.

It's a bit of a misnomer for the base method `Block._createHomogenizedCopy()`, which just performs a `deepcopy` of the block, but that's probably okay. The non-homogenized block will still perform the correct functions, it will just be less efficient than `HexBlock._createHomogenizedCopy()`.

File: armi/reactor/blocks.py (hunk 12 above)
Comment (mgjarrett) [reply armi/reactor/blocks.py]: We spoke offline; updated to handle multiple clad components

File: armi/reactor/converters/uniformMesh.py [hunk 18]
```diff
@@ -680,6 +699,9 @@ def _computeAverageAxialMesh(self):
         refNumPoints = len(src.core.findAllAxialMeshPoints([refAssem])) - 1
         allMeshes = []
         for a in src.core:
+            if a.hasFlags(Flags.CONTROL):
+                # don't include control assemblies; their mesh lines move a bunch
+                continue
```
[the remark this answers is not in the document]
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py]: Removed

File: armi/reactor/converters/uniformMesh.py [hunk 19]
```diff
@@ -307,13 +312,13 @@ def applyStateToOriginal(self):
         runLog.extra(
             f"Parameter remapping time: {completeEndTime-completeStartTime} seconds"
         )
-        self.reset()
```
[the remark this answers is not in the document]
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py:310]: There is a `meshConverter.reset()` call after `applyStateToOriginal` is called here:

https://github.com/terrapower/armi/blob/9d6a405f10a0084b2fea8721f2b6d08c16a0b7a8/armi/physics/neutronics/globalFlux/globalFluxInterface.py#L485

The same pattern exists in the one place I could find `applyStateToOriginal` in internal code.

You could make the argument that it's better to keep the one that's within `applyStateToOriginal` because then we wouldn't ever forget to call it. But I don't see why it's absolutely necessary to run `reset()` immediately after `applyStateToOriginal`. They seem like independent actions, so I think it's okay to keep them separate.

File: armi/reactor/converters/uniformMesh.py [hunk 20]
```diff
@@ -171,6 +176,7 @@ def convert(self, r=None):
                     assem,
                     self.convReactor.core.p.axialMesh[1:],
                     self.blockParamNames,
+                    self.bpm,
```
[the remark this answers is not in the document]
Comment (mgjarrett) [reply armi/reactor/converters/uniformMesh.py]: I did, and I also updated the name to `self.paramMapper` to be more clear, and moved `self.blockParamNames` and `self.reactorParamNames` into the `self.paramMapper`

File: armi/physics/neutronics/parameters.py (hunk 7 above)
Comment (mgjarrett) [reply armi/physics/neutronics/parameters.py:582]: Yes, I think we really only need `rateFis` and `rateProdFis` for the particular use case that I'm thinking of. But there could be other cases where other reaction rates are desired. I didn't know exactly where to draw the line so I just included all of them instead of being picky. But if we're really trying to drill down on performance then we could limit it to just those fission rate parameters.

File: armi/reactor/blocks.py [hunk 21]
```diff
@@ -1519,6 +1535,85 @@ def coords(self, rotationDegreesCCW=0.0):
...
+        by omitting this detailed data and only providing the necessary level of detail for
+        the uniform mesh reactor: number densities on each block.
+
+        .. note: Individual components within a block can have different temperatures, and this
+        can affect cross sections. This temperature variation is captured by the lattice physics
+        module. As long as temperature distribution is correctly captured during cross section
+        generation, it doesn't need to be transferred to the neutronics solver directly through
+        this copy operation.
+
+        .. note: If you make a new block, you must add it to an assembly and a reactor.
```
[the remark this answers is not in the document]
Comment (jakehader) [reply armi/reactor/blocks.py:1564]: I think it does!

File: armi/reactor/blocks.py (hunk 1 above)
Comment (john-science) [reply armi/reactor/blocks.py]: Yeah, we will need to find a way to fix this.  I don't like the idea that we are forcing the uniform mesh to only work for hex assemblies now.  (Because this is used pretty high-level in `makeAssemWithUniformMesh()`).

File: armi/reactor/blocks.py [hunk 22]
```diff
@@ -1570,6 +1570,7 @@ def _createHomogenizedCopy(self, pinSpatialLocators=False):
 
         b = self.__class__(self.getName(), height=self.getHeight())
         b.setType(self.getType())
+        b.p.flags = self.p.flags
```
Comment (keckler) [inline armi/reactor/blocks.py:1573]: I think setting the flags from within the `setType()` call itself would be more efficient, because it would avoid the internal call to `Flags.fromStringIgnoreErrors(typ)`.

https://github.com/terrapower/armi/blob/main/armi/reactor/composites.py#L757-L784

Comment (keckler) [reply armi/reactor/blocks.py:1573]: But either should work. I'm testing the way that you have it right now.

Comment (mgjarrett) [reply armi/reactor/blocks.py]: Updated!

File: armi/reactor/blocks.py (hunk 1 above)
Comment (mgjarrett) [reply armi/reactor/blocks.py]: It still works with non-hex assemblies, it just uses `copy.deepcopy(block)`. It's the same behavior that existed before, which is less efficient but still works.

**Comment by keckler** [comment]
Just putting this here to help back up this PR.

I've been testing locally with the version at the following hash: https://github.com/terrapower/armi/pull/1042/commits/81f1a81cb7891987777c522f0c49171fbce2e69f

This appears to speed up my calculation _a lot_. A calculation which previously took nearly 5 days is on track now to take only about 14 hrs.

A little bit of information about that run to help put that in context:
- I am using DIF3D/VARIANT P1P0 and MC2, so not the heaviest physics kernels
- I am doing critical control rod iterations at every step
- Anywhere between 1-4 depletion steps at each cycle, for nearly 50 cycles
- No other physics
- `detailedAxialExpansion: True`

This run was previously spending a lot of time doing uniform mesh conversions. That time has been reduced substantially, so the overall calculation time is now reduced by a factor of ~8-10.

Awesome!!

**Review by keckler — COMMENTED** [review]
The changes _look_ good to me. I am running a test case and will comment again with the results of that.

**Comment by keckler** [comment]
My test case worked well here. No complaints from me!

**Review by jakehader — APPROVED** [review]
Looks like this is functional and I propose to merge as soon as possible if @keckler and @john-science agree.

File: armi/reactor/blocks.py [hunk 23]
```diff
@@ -1518,6 +1534,87 @@ def coords(self, rotationDegreesCCW=0.0):
...
+        b.setType(self.getType(), self.p.flags)
+
+        # assign macros and LFP
+        b.macros = self.macros
+        b._lumpedFissionProducts = self._lumpedFissionProducts
+        b.p.buGroup = self.p.buGroup
+
+        hexComponent = Hexagon(
+            "homogenizedHex",
+            "_Mixture",
```
Comment (keckler) [inline armi/reactor/blocks.py:1580]: @mgjarrett Can you help me understand why this new `_Mixture` material is needed here? I see that it is just a subclass of `Material` with no differences. Couldn't we have just built a Hexagon with `Material` instead?
